File: src/main/java/typeSpecifications/querySpecification/QueryNode.java
package typeSpecifications.querySpecification;


import java.util.LinkedList;

// class for specifying a query node.
// the node has an identifier and is possibly an initialState or a finalState.
public class QueryNode {

    private String identifier;
    private boolean initialState;
    private boolean finalState;
    private LinkedList<QueryEdge> edges;

    public QueryNode(String identifier, boolean initialState, boolean finalState) {
        this.identifier = identifier;
        this.initialState = initialState;
        this.finalState = finalState;
        edges = new LinkedList<>();
    }


    public boolean isInitialState() {
        return initialState;
    }

    public boolean isFinalState() {
        return finalState;
    }

    public String getIdentifier() {
        return identifier;
    }

    public LinkedList<QueryEdge> getEdges() {
        return edges;
    }

    public boolean equals(QueryNode otherNode) {
        // compare basic parameters
        if (equalsExcludingEdges(otherNode)) {
            // compare edges
            {
                if (this.edges == null || otherNode.edges == null) {
                    return false;
                }

                if (this.edges.size() != otherNode.edges.size()) {
                    return false;
                }

                return equalsOtherEdgeSet(otherNode.edges);
            }
        }

        return false;
    }

    private boolean equalsOtherEdgeSet(LinkedList<QueryEdge> otherEdgeList) {
        int pairsFound = 0;
        for (QueryEdge thisEdge : this.edges) {
            for (QueryEdge otherEdge : otherEdgeList) {
                if (thisEdge.equals(otherEdge)) {
                    pairsFound++;
                    break;
                }
            }
        }
        return pairsFound == (this.edges.size());
    }

    public boolean equalsExcludingEdges(QueryNode otherNode) {
        return this.initialState == otherNode.initialState && this.finalState == otherNode.finalState && this.identifier.equals(otherNode.identifier);
    }
}

File: tin v2/src/main/kotlin/tin/model/query/QueryNode.kt
package tin.model.query

import java.util.*

class QueryNode(
        val identifier: String,
        val isInitialState: Boolean,
        val isFinalState: Boolean,
        val edges: LinkedList<QueryEdge>
) {
    fun equals(otherNode: QueryNode): Boolean {
        // compare basic parameters
        if (equalsExcludingEdges(otherNode)) {
            // compare edges
            run {
                return if (this.edges.size != otherNode.edges.size) {
                    false
                } else equalsOtherEdgeSet(otherNode.edges)
            }
        }
        return false
    }

    private fun equalsOtherEdgeSet(otherEdgeList: LinkedList<QueryEdge>?): Boolean {
        var pairsFound = 0
        for (thisEdge in edges!!) {
            for (otherEdge in otherEdgeList!!) {
                if (thisEdge.equals(otherEdge)) {
                    pairsFound++
                    break
                }
            }
        }
        return pairsFound == edges.size
    }

    fun equalsExcludingEdges(otherNode: QueryNode): Boolean {
        return isInitialState == otherNode.isInitialState && isFinalState == otherNode.isFinalState && identifier == otherNode.identifier
    }
}
